stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -462,61 +462,33 @@ norm_sqr(x::Union{T,Complex{T},Rational{T}}) where {T<:Integer} = abs2(float(x))
 
 function generic_norm2(x)
     maxabs = normInf(x)
-    (maxabs == 0 || isinf(maxabs)) && return maxabs
-    (v, s) = iterate(x)::Tuple
+    (ismissing(maxabs) || maxabs == 0 || isinf(maxabs)) && return maxabs
     T = typeof(maxabs)
-    if isfinite(length(x)*maxabs*maxabs) && maxabs*maxabs != 0 # Scaling not necessary
-        sum::promote_type(Float64, T) = norm_sqr(v)
-        while true
-            y = iterate(x, s)
-            y === nothing && break
-            (v, s) = y
-            sum += norm_sqr(v)
-        end
-        return convert(T, sqrt(sum))
+    sT = promote_type(Float64, T)
+    maxabs_st = sT(maxabs)
+    if isfinite(length(x)*maxabs_st*maxabs_st) && maxabs_st*maxabs_st != 0 # Scaling not necessary
+        return convert(T, sqrt(mapreduce(v -> sT(norm_sqr(v)), +, x)))
     else
-        sum = abs2(norm(v)/maxabs)
-        while true
-            y = iterate(x, s)
-            y === nothing && break
-            (v, s) = y
-            sum += (norm(v)/maxabs)^2
-        end
-        return convert(T, maxabs*sqrt(sum))
+        invmaxabs = inv(maxabs_st)
+        return convert(T, maxabs*sqrt(mapreduce(v -> abs2(sT(norm(v))*invmaxabs), +, x)))
     end
 end
 
 # Compute L_p norm ‖x‖ₚ = sum(abs(x).^p)^(1/p)
 # (Not technically a "norm" for p < 1.)
 function generic_normp(x, p)
-    (v, s) = iterate(x)::Tuple
+    T = typeof(float(first(x)))
```

Comment:
If `x` is an iterable/array of arrays, then this will not give the desired result, right? Maybe it should be
```suggestion
    T = typeof(float(norm(first(x))))
```
?